Sources/CartonHelpers/DiagnosticsParser.swift

```diff
@@ -0,0 +1,207 @@
+  func makeBuilder() -> TerminalOutputBuilder {
+    .init()
+  }
+  
+  struct TerminalOutputBuilder: OutputBuilder {
+    var output: String = ""
+    
+    mutating func addToken(_ token: String, ofType type: TokenType) {
+      output.append("\(token, color: type.color)")
+    }
+    
+    mutating func addPlainText(_ text: String) {
+      output.append(text)
+    }
+    
+    mutating func addWhitespace(_ whitespace: String) {
+      output.append(whitespace)
+    }
+    
+    mutating func build() -> String {
+      output
+    }
+  }
+}
+
+/// Parses and re-formats diagnostics output by the Swift compiler.
+///
+/// The compiler output often repeats iteself, and the diagnostics can sometimes be difficult to read.
+/// This reformats them to a more readable output.
+struct DiagnosticsParser {
+  enum Regex {
+    /// The output has moved to a new file
+    static let enterFile = try! NSRegularExpression(pattern: #"\[\d+\/\d+\] Compiling \w+ "#)
+    /// A message is beginning with the line # following the `:`
+    static let line = try! NSRegularExpression(pattern: #"(\/\w+)+\.\w+:"#)
+  }
+  
+  struct CustomDiagnostic {
+    let kind: Kind
+    let line: String.SubSequence
+    let char: String.SubSequence
+    let code: String
+    let message: String.SubSequence
+    
+    enum Kind: String {
+      case error, warning, note
+      var color: String {
+        switch self {
+        case .error: return "[41;1m"
+        case .warning: return "[43;1m"
+        case .note: return "[47;1m"
+        }
+      }
+    }
+  }
+  
+  fileprivate static let highlighter = SyntaxHighlighter(format: TerminalOutputFormat())
+  
+  func parse(_ output: String, _ terminal: InteractiveWriter) {
+    let lines = output.split(separator: "\n")
+    var lineIdx = 0
+    
+    var diagnostics = [String.SubSequence:[CustomDiagnostic]]()
+    
+    var currFile: String.SubSequence?
+    var fileMessages = [CustomDiagnostic]()
+    
+    while lineIdx < lines.count {
+      let line = lines[lineIdx]
+      if let file = line.matches(regex: Regex.enterFile) {
+        if let currFile = currFile {
+          diagnostics[currFile] = fileMessages
+        }
+        currFile = file
+        fileMessages = []
+      } else if currFile != nil {
+        if let message = line.matches(regex: Regex.line) {
+          let components = message.split(separator: ":")
+          if components.count > 3 {
+            lineIdx += 1
+            fileMessages.append(
+              .init(
+                kind: CustomDiagnostic.Kind(rawValue: String(components[2].trimmingCharacters(in: .whitespaces))) ?? .note,
+                line: components[0],
+                char: components[1],
+                code: String(lines[lineIdx]),
+                message: components[3]
+              )
+            )
+          }
+        }
+      } else {
+        terminal.write(String(line) + "\n", inColor: .cyan)
+      }
+      lineIdx += 1
+    }
+    
```

Comment:
<!--
  1 failure:  Lines should not ...
  0 warning: 
  
  
  DangerID: danger-id-Danger;
  File: Sources/CartonHelpers/DiagnosticsParser.swift;
  Line: 147;
-->
- :no_entry_sign: Lines should not have trailing whitespace. (`trailing_whitespace`)